File: app/src/main/java/com/komica/reader/adapter/BoardCategoryAdapter.java
package com.komica.reader.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.komica.reader.R;
import com.komica.reader.model.Board;
import com.komica.reader.model.BoardCategory;
import java.util.List;

public class BoardCategoryAdapter extends RecyclerView.Adapter<BoardCategoryAdapter.CategoryViewHolder> {

    private List<BoardCategory> categories;
    private OnBoardClickListener onBoardClickListener;
    private BoardAdapter.OnFavoriteClickListener onFavoriteClickListener;

    public interface OnBoardClickListener {
        void onBoardClick(Board board);
    }

    public BoardCategoryAdapter(List<BoardCategory> categories, OnBoardClickListener listener, BoardAdapter.OnFavoriteClickListener favoriteListener) {
        this.categories = categories;
        this.onBoardClickListener = listener;
        this.onFavoriteClickListener = favoriteListener;
    }

    @NonNull
    @Override
    public CategoryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_board_category, parent, false);
        return new CategoryViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CategoryViewHolder holder, int position) {
        BoardCategory category = categories.get(position);
        holder.bind(category, position);
    }

    @Override
    public int getItemCount() {
        return categories.size();
    }

    class CategoryViewHolder extends RecyclerView.ViewHolder {
        private View categoryHeader;
        private TextView categoryName;
        private ImageView expandIcon;
        private RecyclerView boardsRecyclerView;

        public CategoryViewHolder(@NonNull View itemView) {
            super(itemView);
            categoryHeader = itemView.findViewById(R.id.categoryHeader);
            categoryName = itemView.findViewById(R.id.categoryName);
            expandIcon = itemView.findViewById(R.id.expandIcon);
            boardsRecyclerView = itemView.findViewById(R.id.boardsRecyclerView);
            boardsRecyclerView.setLayoutManager(new LinearLayoutManager(itemView.getContext()));
        }

        public void bind(BoardCategory category, int position) {
            categoryName.setText(category.getName());
            
            categoryHeader.setOnClickListener(v -> {
                category.setExpanded(!category.isExpanded());
                notifyItemChanged(position);
            });

            if (category.isExpanded()) {
                expandIcon.setImageResource(R.drawable.ic_expand_less);
                boardsRecyclerView.setVisibility(View.VISIBLE);
            } else {
                expandIcon.setImageResource(R.drawable.ic_expand_more);
                boardsRecyclerView.setVisibility(View.GONE);
            }

            BoardAdapter boardAdapter = (BoardAdapter) boardsRecyclerView.getAdapter();
            if (boardAdapter == null) {
                boardAdapter = new BoardAdapter(category.getBoards(), onBoardClickListener, onFavoriteClickListener);
                boardsRecyclerView.setAdapter(boardAdapter);
            } else {
                boardAdapter.updateData(category.getBoards());
            }
        }
    }
}

File: app/src/main/java/com/komica/reader/adapter/BoardCategoryAdapter.kt
package com.komica.reader.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.komica.reader.R
import com.komica.reader.model.Board
import com.komica.reader.model.BoardCategory

class BoardCategoryAdapter(
    private var categories: List<BoardCategory>,
    private val onBoardClickListener: OnBoardClickListener?,
    private val onFavoriteClickListener: BoardAdapter.OnFavoriteClickListener?
) : RecyclerView.Adapter<BoardCategoryAdapter.CategoryViewHolder>() {

    fun interface OnBoardClickListener {
        fun onBoardClick(board: Board)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_board_category, parent, false)
        return CategoryViewHolder(view)
    }

    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
        holder.bind(categories[position], position)
    }

    override fun getItemCount(): Int = categories.size

    inner class CategoryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val categoryHeader: View = itemView.findViewById(R.id.categoryHeader)
        private val categoryName: TextView = itemView.findViewById(R.id.categoryName)
        private val expandIcon: ImageView = itemView.findViewById(R.id.expandIcon)
        private val boardsRecyclerView: RecyclerView = itemView.findViewById(R.id.boardsRecyclerView)

        init {
            boardsRecyclerView.layoutManager = LinearLayoutManager(itemView.context)
        }

        fun bind(category: BoardCategory, position: Int) {
            categoryName.text = category.name

            categoryHeader.setOnClickListener {
                category.isExpanded = !category.isExpanded
                notifyItemChanged(position)
            }

            if (category.isExpanded) {
                expandIcon.setImageResource(R.drawable.ic_expand_less)
                boardsRecyclerView.visibility = View.VISIBLE
            } else {
                expandIcon.setImageResource(R.drawable.ic_expand_more)
                boardsRecyclerView.visibility = View.GONE
            }

            val boardAdapter = boardsRecyclerView.adapter as? BoardAdapter
            if (boardAdapter == null) {
                boardsRecyclerView.adapter = BoardAdapter(
                    category.boards,
                    onBoardClickListener,
                    onFavoriteClickListener
                )
            } else {
                boardAdapter.updateData(category.boards)
            }
        }
    }
}
